File: app/src/main/java/com/jnjcomu/edison/factory/UserFactory.java
package com.jnjcomu.edison.factory;

import android.support.annotation.NonNull;

import com.jnjcomu.edison.model.Ticket;
import com.jnjcomu.edison.model.User;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;

/**
 * @author CodeRi13 <ruto1924@gmail.com>
 * @since 2017-05-24
 */

public class UserFactory {

    @NonNull
    public static User extractUser(Ticket ticket) {
        String token = ticket.getTicketCode();
        Claims claims = Jwts.parser()
                .parseClaimsJws(token)
                .getBody();

        int userID = claims.get("id", Integer.class);
        int userSerial = claims.get("serial", Integer.class);
        String userName = claims.get("name", String.class);
        String userType = claims.get("userType", String.class);

        return new User(
                userID,
                userName,
                extractGradeFromSerial(userSerial),
                extractClazzFromSerial(userSerial),
                extractNumberFromSerial(userSerial)
        );
    }

    public static int extractGradeFromSerial(int serial) {
        return serial / 1000;
    }

    public static int extractClazzFromSerial(int serial) {
        return (serial / 100) % 10;
    }

    public static int extractNumberFromSerial(int serial) {
        return serial - (serial / 100 * 100);
    }
}

File: app/src/main/java/com/jnjcomu/edison/factory/UserFactory.kt
package com.jnjcomu.edison.factory

import com.jnjcomu.edison.model.Ticket
import com.jnjcomu.edison.model.User

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts

/**
 * @author CodeRi13 <ruto1924@gmail.com>
 * @since 2017-05-24
 */

object UserFactory {

    fun extractUser(ticket: Ticket): User {
        val token = ticket.ticketCode
        val claims = Jwts.parser()
                .parseClaimsJws(token)
                .body

        val userID = claims.get("id", Int::class.java)
        val userSerial = claims.get("serial", Int::class.java)
        val userName = claims.get("name", String::class.java)
        val userType = claims.get("userType", String::class.java)

        return User(
                userID,
                userName,
                extractGradeFromSerial(userSerial),
                extractClazzFromSerial(userSerial),
                extractNumberFromSerial(userSerial)
        )
    }

    fun extractGradeFromSerial(serial: Int): Int {
        return serial / 1000
    }

    fun extractClazzFromSerial(serial: Int): Int {
        return serial / 100 % 10
    }

    fun extractNumberFromSerial(serial: Int): Int {
        return serial - serial / 100 * 100
    }
}
